### ISSUE TITLE
Will never work as intended

### ISSUE THREAD

**Issue by ugexe** [ufobat/p6-ResourceExtractor#1]
FWIW this will never work as intended. You cannot just copy files around on the filesystem without mangling them because not all filesystems accept the same path names for a given directory. If modules start using this then the entire cross-system compatibility rakudo has tried to maintain will start to crumble away as uninformed people start/continue to use solutions like this. Also this won't work for non-installed modules -- something `%?RESOURCES` current implementation already does.

**Comment by ufobat** [comment]
I fully agree to your first point. The whole module felt like a bad hack to me as well. And I wanted to but a big **WARNING** into the README.md. But then, where is the point in this module?!

Has the access to `%?RESOURCES`  of non installed modules recently been added? I wasn't aware that this actualyl works! Great!

But I am still not willing to give this idea up.
First, please don't take the name of this module literal. I am still trying to figure out what is reasonable.
Second, and that's why I am not willing to give it up. I am still unhappy about the features of `%?RESOURCES`. You can not "introspect" it. 
* Afaik there is no way of doing something like this `for %?RESOURCES.kv -> $name, $io { ... }`
* If this was possible you could still not list all files of a certain folder that was stored in the  `resources/` dir. Similar to `for %?RESOURCES<folder-foo>.kv -> $name, $io { ... }`

Without those features _understanding_ `%*RESOURCES`, like @b2gills (Brat Gilbert) asked Bailador and Cro to do, is pretty inconvinient.

in 7b1421d855d0dc68e70827219c8001f861bad44e I tried to seperate the reasonable code from the bad code.

Today pmurias and I had the idea to implement a `IO::Path'ish` interface to the resources so it is still all in memory and doen't require special mangling, I am not sure if this is a good idea, because `role IO` doesnt describe an interface and `IO::Path` is already concrete and strictly filesystem related, there is nothing in the middle? 
Exporting in a tempdir() might still be part of this module, but would/should never be used in oder to provide access to the recources like I wanted to do before, for my Cro/Bailador applications.

**Comment by ugexe** [comment]
`%?RESOURCES` has always worked for non-installed modules. Both `zef` and `Inline::Perl5` have used it since at least `6.c`. I would argue @b2gills is spot on -- it is something to be implemented by whatever is using it. 

Consider: how would a theoretical `IO::Resources` work with:
* https://github.com/ugexe/Perl6-CompUnit--Repository--Github
* https://github.com/ugexe/Perl6-CompUnit--Repository--Tar

Those `CUR`s have not accessed via a conventional filesystem ( e.g. loading a module from a tar file does not extract any files from the archive, it outputs the data to stdout and captures that to load from bytes ). Such a solution requires `IO::Resources` to have specially coded code paths for every known `CompUnit::Repository` in existence ( not possible ). Instead it is up to the `Distribution` ( really the `CompUnit::Repository` but that is an artifact of the original design ) to be the `IO` interface to a distribution via its meta data. In this way any special name mangling each `CompUnit::Repository` might use can handle the translation of `%?RESOURCES`.

Alternatively don't use `%?RESOURCES`. There is already a way to get the keys of `%?RESOURCES` and that is through the META data interface: `Distribution`. `say $dist.content($_) for $dist.meta<resources>.list`

```
    # Only works for installed modules until/if this pr gets merged:
    # https://github.com/rakudo/rakudo/pull/1812

    my $spec = CompUnit::DependencySpecification.new(:short-name<Zef::Client>);
    my $dist = $*REPO.need($spec).distribution;
    say $dist.content($_) for $dist.meta<resources>.list;
```

Is this really that unreasonable compared to duplicating this functionality and increasing complexity ( which implements shim functions of nearly all `IO` methods already ) of `%?RESOURCES`?

---

With all of that being said -- if you want to be a hero and do this in a better way then implement `$*DISTRIBUTION` such that inside your code you could do e.g. `$*DISTRIBUTION.content($*DISTRIBUTION.meta<resources>.head)` instead of having to lookup by name. With that you could create an `IO::Resources` to simplify things a bit while also working with any type of future `CUR`.

**Comment by ugexe** [comment]
As for `%?RESOURCES<folder-foo>.kv -> $name, $io { ... }`: just split on `/`. Using a forward slash separator is the convention used elsewhere in the metadata. This makes it trivial to parse properly while being consistent with the rest of the meta data which does *not* e.g. `%provides<dir><subdir>`. So internally you'll likely not see this change, but I suppose it could be something that could go in a module ( although its so simple to split a string I'd argue its not as useful as it may seem ).
